ocaml/xapi/helpers.ml

```diff
@@ -1955,6 +1955,19 @@ end = struct
     Xapi_psr_util.load_psr_pool_secrets ()
 end
 
+let ( let@ ) f x = f x
+
+let with_temp_fd fd f = finally (fun () -> f fd) (fun () -> Unix.close fd)
+
+let with_temp_file ?mode prefix suffix f =
+  let temp_file, out_ch = Filename.open_temp_file ?mode prefix suffix in
+  finally (fun () -> f (temp_file, out_ch)) (fun () -> Unix.unlink temp_file)
+
+let with_temp_fd_of_temp_file ?mode prefix suffix f =
+  let@ temp_file, out_ch = with_temp_file ?mode prefix suffix in
+  let@ temp_out_fd = out_ch |> Unix.descr_of_out_channel |> with_temp_fd in
```

Comment:
This will cause a warning to be printed to stderr when Sys.enable_runtime_warnings is used:
```
utop # Sys.enable_runtime_warnings true;;
- : unit = ()
─( 10:13:35 )─< command 3 >───────────────────────────────────────────────────────────────────────────────────────────────────────────────────{ counter: 0 }─
utop # Filename.open_temp_file "foo" "foo" |> snd |> Unix.descr_of_out_channel |> Unix.close;;
- : unit = ()
[ocaml] (use Sys.enable_runtime_warnings to control these warnings)
[ocaml] channel opened on file '/tmp/foo7d4427foo' dies without being closed
```

We should really try to close the channel instead of the FD it contains inside.